src/Components/Components/src/EventHandlerInvokerFactory.cs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components
+{
+    public sealed class EventHandlerInvokerFactory
+    {
+        public Action CreateDelegate(object receiver, Action action)
+        {
+            if (receiver == null)
+            {
+                throw new ArgumentNullException(nameof(receiver));
+            }
+
+            if (action == null)
+            {
+                throw new ArgumentNullException(nameof(action));
+            }
+
+            // Determine whether the action is already a bound-delegate attached to handler.
+            //
+            // This is important for a case like the following:
+            //
+            //  @* in Outer.cshtml *@
+            //  <Inner OnClick="@OnClick" />
+            //  @functions {
+            //      void OnClick() { ... }
+            //  }
+            //
+            //  @* in Inner.cshtml *@
+            //  <button onclick="@OnClick">Click Me!</button>
+            //  @functions {
+            //      [Parameter] Action OnClick { get; set; }
+            //  }
+            //
+            // We want to make sure to call StateHasChanged on the Outer component instead
+            // of just the Inner component. We can't rely on action.Target to always point
+            // to the outer component because in the case of a non-capturing lambda it won't.
```

Comment:
Nit: it's the capturing (not non-capturing) case where delegate.Target won't be `Outer`.

Reply:
I think maybe we're just disagreeing on terminology maybe?

** Capturing Lambda**

```html
<MdButton OnClicked="@(() => DoSomething())" />

@functions {
    void DoSomething() { ... }
}
```


**Non-capturing Lambda**

```html
<MdButton OnClicked="@(() => Console.Log("YEET"))" />
```


**Bound Delegate**

```html
<MdButton OnClicked="DoSomething" />

@functions {
    void DoSomething() { ... }
}
```

Reply:
OK yeah. I was distinguishing based on whether the compiler would generate a hidden class to hold captured variables, which it does if and only if you are capturing something *other than* `this`, e.g.:

```html
@foreach(var thing in myCollection)
{
    <MdButton OnClicked="@(() => DoSomething(thing))" />
}
```

In this case, the delegate's `Target` will be an instance of the compiler-generated class instead of the Razor Component.